### PR TITLE
Scientific parser

### PR DESCRIPTION
**Pull request by rnjtranjan**
https://github.com/composewell/streamly/issues/2276

### CODE REVIEW

File: src/Streamly/Internal/Data/Unicode/Parser/Extra.hs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+-- >>> scientificParser "31e+3"
+-- Right "31000.0"
+--
+-- >>> scientificParser "-3.1e4"
+-- Right "-31000.0"
+--
+-- >>> scientificParser "-3.1e-4"
+-- Right "-0.00031"
+scientific :: (Monad m) => Parser Char m Scientific
+scientific = Parser step initial extract
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Unicode/Parser/Extra.hs:65]: Add an INLINE.

**Comment by rnjtranjan** [comment]
Benchmarks:
#No INLINE
scientific:                                   OK (0.16s)
      11.1 ms ± 1.1 ms
	
#With INLINE	
scientific:                                   OK (0.27s)
      8.70 ms ± 573 μs

File: src/Streamly/Internal/Data/Unicode/Parser/Extra.hs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    where
+
+    intToInteger :: Int -> Integer
+    intToInteger = fromIntegral
+
+    combineNum buf num = buf * 10 + num
+
+    initial = pure $ IPartial SPInitial
```
Comment (adithyaov) [inline src/Streamly/Internal/Data/Unicode/Parser/Extra.hs:75]: Need to add proper `Inline` statements to `step`, `initial`, and `extract`.

File: streamly.cabal [hunk 3]
```diff
@@ -401,6 +401,9 @@ library
                      , Streamly.Network.Socket
                      , Streamly.Network.Inet.TCP
 
+                    -- Unicode parser
+                     , Streamly.Internal.Data.Unicode.Parser.Extra
```
Comment (adithyaov) [inline streamly.cabal:405]: Should be `Streamly.Internal.Unicode.Parser.Extra`

File: test/Streamly/Test/Data/Parser.hs [hunk 4]
```diff
@@ -1129,6 +1132,29 @@ takeStartBy_ =
...
+scientificFixFP :: Property
+scientificFixFP =  forAll (chooseDouble (-0.00099, 123445.67998)) $ \ls ->
+                case runIdentity $ S.parse parser (S.fromList (show ls)) of
+                     Right val -> property (val == show ls)
+                     Left _ -> property False
+
+                    where
+                    formatter = Scientific.formatScientific Scientific.Fixed Nothing
+                    parser = formatter <$> Parser.scientific
+
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Parser.hs:1157]: Use proper indentation.

File: test/Streamly/Test/Data/Parser.hs [hunk 5]
```diff
@@ -1129,6 +1132,29 @@ takeStartBy_ =
             where
                 predicate = odd
                 parser = P.takeStartBy_ predicate FL.toList
+
+-- Scientific notation
+scientificExpFP :: Property
+scientificExpFP =  forAll (chooseDouble (-99.99e-12, 1234.4567e+234)) $ \ls ->
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Parser.hs:1138]: Can you explain why you chose this range?

File: test/Streamly/Test/Data/Parser.hs [hunk 6]
```diff
@@ -1129,6 +1132,29 @@ takeStartBy_ =
...
+                     Left _ -> property False
+
+                    where
+                    formatter = Scientific.formatScientific Scientific.Exponent Nothing
+                    parser = formatter <$> Parser.scientific
+
+-- Standard decimal notation.
+scientificFixFP :: Property
+scientificFixFP =  forAll (chooseDouble (-0.00099, 123445.67998)) $ \ls ->
+                case runIdentity $ S.parse parser (S.fromList (show ls)) of
```
Comment (adithyaov) [inline test/Streamly/Test/Data/Parser.hs:1150]: Can you explain why you chose this range?

File: benchmark/Streamly/Benchmark/Data/Parser.hs [hunk 7]
```diff
@@ -628,6 +630,27 @@ choice value =
         (PR.choice (replicate value (PR.satisfy (< 0))) AP.<|> PR.satisfy (> 0))
 -}
 
+{-# INLINE sourceUnfoldrMc #-}
+sourceUnfoldrMc :: Monad m => Int -> Int -> Stream m Char
```
Comment (adithyaov) [inline benchmark/Streamly/Benchmark/Data/Parser.hs:634]: Can the suffix `c` be capital?
Also why do we need it?
Can't we just use `sourceUnfoldrM`? and map `chr` on it?

File: benchmark/Streamly/Benchmark/Data/Parser.hs [hunk 8]
```diff
@@ -628,6 +630,27 @@ choice value =
...
+sourceUnfoldrMc value n = Stream.unfoldrM step n
+    where
+    step cnt =
+        if cnt > n + value
+        then return Nothing
+        else return (Just ('1', cnt + 1))
+
+-- | Takes a fold method, and uses it with a default source.
+{-# INLINE benchIOSinkc #-}
+benchIOSinkc
```
Comment (adithyaov) [inline benchmark/Streamly/Benchmark/Data/Parser.hs:644]: You probably don't need this.

File: test/Streamly/Test/Data/Parser.hs (hunk 5 above)
Comment (rnjtranjan) [reply test/Streamly/Test/Data/Parser.hs:1138]: cover the value range of negative and positive with a negative and positive exponent.

File: test/Streamly/Test/Data/Parser.hs (hunk 6 above)
Comment (rnjtranjan) [reply test/Streamly/Test/Data/Parser.hs:1150]: cover the value range of negative and positive with a negative and positive exponent.

**Comment by harendra-kumar** [comment]
We can keep the tests for scientific and remove the implementation. Use the number parser for tests.

**Comment by harendra-kumar** [comment]
The `number` parser covers this functionality.

**Pull request closed without merging**